### DISCUSSION TITLE
Migration - What's changed? V1 -> V2

### DISCUSSION THREAD

**Discussion by PhillypHenning** [bitovi/bitops#215] [General]
This discussion will be used to capture any changes that will need to be noted in the migration documentation.

**Comment by PhillypHenning** [comment]
BitOps core 🧵

**Reply by PhillypHenning** [reply]
Bitops is no longer using the `ENVIRONMENT` value, it instead uses; `BITOPS_ENVIRONMENT`

Easiest migration solution; 
Update startup command to read as followed;
```
docker run \
-e BITOPS_ENVIRONMENT=$ENVIRONMENT   \    <-- If the ENVIRONMENT variable is being used. This is a simple approach. 
-v /path/to/ops_repo:/opt/bitops_deployment    \
bitovi/bitops:latest
```

**Reply by PhillypHenning** [reply]
BitOps core exported environment variables now have a prefix of `BITOPS_`

BitOps plugins (tool specific) environment variable exports are prefixed with; `BITOPS_<PLUGIN NAME>`

**Examples**
BITOPS_TERRAFORM_ENV_VAR
BITOPS_ANSIBLE_ENV_VAR

**Reply by PhillypHenning** [reply]
# Value add
ENVIRONMENT is a contentious variable name so to avoid unnecessary conflict we've decided to prefix it with `BITOPS_`.

This is also a pattern we use for exported environment variables.

**Comment by PhillypHenning** [comment]
OpsRepo 🧵

**Reply by PhillypHenning** [reply]
ops_repo level bitops.config.yaml have had one important update; The cli attribute "stack-action" has been added. This attribute is used to tell the bitops plugin which method it is invoking. 

For example, the terraform plugin, has 3 stack-actions;
- apply
- plan
- destroy

**Old method**
```
terraform:
    cli: {}
    options:
        command: apply
```

**New method**
```
terraform:
  cli:
    stack-action: apply
```

**Reasoning for the change**
This pattern is used by BitOps to standardize how a plugin specifies an action

**Reply by mickmcgrath13** [reply]
@PhillypHenning @shyamrayaprolu should it be:
```
terraform:
  options:
    stack-action: apply
```

the reason being `${tool}.cli` should be reserved for the paramegers the cli tool itself uses, and `${tool}.options` should be for what bitops uses to control the cli tool itself.

thoughts?

**Reply by PhillypHenning** [reply]
Personally I'd like to keep it in the CLI section as; 
- It does directly related to what the CLI is about to act on

As well, it's not optional, and though 'options' doesn't necessarily mean optional, I do see them as somewhat related. (There will be circumstance when that isn't the case.

**Reply by PhillypHenning** [reply]
# Value add
As we move into the world of plugins, we needed a way to consolidate the "actions" that a plugin can take. The `stack-action` attribute creates a clear pattern for plugins to follow and for BitOps to process.

**Comment by PhillypHenning** [comment]
Plugin 🧵

**Reply by PhillypHenning** [reply]
@shyamrayaprolu please add a "stack actions" section where relevant to the top of the plugin READMEs

**Reply by PhillypHenning** [reply]
**Ansible**
*changed attributes*

*depreciated attributes*
- ansible.cli.vault_id
- ansible.cli.vault_password
- ansible.options.verbosity

*new attributes*

<hr/>

**Terraform**
*changed attributes*

*depreciated attributes*

*new attributes*
- ansible.options.init-upgrade

<hr/>

**Helm**
*changed attributes*
- helm.options.uninstall-charts --> Changed to --> helm.options.uninstall
- helm.option.kubeconfig.fetch.enabled --> Changed to --> helm.option.kubeconfig.fetch.fetch_kubeconfig

*depreciated attributes*


*new attributes*
- helm.options.default-root-dir
- helm.options.default-dir-flag
- helm.options.default-sub-dir

<hr/>

**Cloudformation**
*changed attributes*
- cloudformation.options.cfn-files.parameters.enabled --> Changed to --> cloudformation.options.cfn-files.parameters.template-param-flag
- cloudformation.options.cfn-files.parameters.template-file --> Changed to --> cloudformation.options.cfn-files.parameters.template-param-file

*depreciated attributes*


*new attributes*
- cloudformation.options.s3bucket
- cloudformation.options.s3prefix

**Reply by PhillypHenning** [reply]
# Value add
misc. 

A few new options were added, some were modified to use new patterns, and a very few were depreciated.

(Phil note) I don't foresee this being an issue)

**Comment by PhillypHenning** [comment]
[Docs](https://github.com/bitovi/bitops/pull/225) have been added capturing this

**Comment by mickmcgrath13** [comment]
https://bitovi.github.io/bitops/migration/2.0.0-migration/
